### PR TITLE
implement flake8-type-checking

### PR DESCRIPTION
**Pull request by half-duplex**
### Description
Tin. See #1765

### Checklist
- [x] I have read [CONTRIBUTING.md](https://github.com/sopel-irc/sopel/blob/master/CONTRIBUTING.md)
- [x] I can and do license this contribution under the EFLv2
- [x] No issues are reported by `make qa` (runs `make quality` and `make test`)
- [x] I have tested the functionality of the things this change touches

### CODE REVIEW

**Comment by half-duplex** [comment]
Had to restrict it to python 3.8+ since flake8-type-checking doesn't support 3.7...

**Comment by dgw** [comment]
Think I threw it on the todo pile in #1765 instead of just doing it immediately _because_ there's no support for 3.7… But there's actually nothing that breaks if someone runs style checks on 3.7, only the potential for missing TC error codes that CI will flag. And probably no one should be running their main development on anything older than 3.8 or 3.9… so I declare my past self silly.

Can you exclude/ignore the TC200 error series? We want to use delayed evaluation (it's why our files use `from __future__ import annotations`), not forward references. Might also want `type-checking-exempt-modules = typing`, which would make the places where `from typing` appears _both_ inside and outside the `TYPE_CHECKING` block unnecessary.

(My present self is also silly, just for different reasons.)

**Comment by dgw** [comment]
Cool. Time to ask for review from the one with actual experience using mypy.

**Comment by Exirel** [comment]
That's interesting. Maybe it's also time to officially add `mypy` to the mix then?

**Comment by dgw** [comment]
Could be, if it doesn't generate any errors after merging with `master`. Whatcha think about this otherwise?

**Review by Exirel — APPROVED** [review]
I'm OK with that as long as ` make mypy` works.

**Review by Exirel — CHANGES_REQUESTED** [review]

File: setup.cfg [hunk 1]
```diff
@@ -53,6 +54,7 @@ ignore =
     # These would require future imports that are not needed any more on Sopel's
     # oldest supported Python version (3.7).
     FI10,FI11,FI12,FI13,FI14,FI15,FI16,FI17,
+    TC2,  # We use postponed annotation evaluation
```
[the remark this answers is not in the document]
Comment (half-duplex) [reply setup.cfg]: Fixed

**Review by Exirel — APPROVED** [review]

**Pull request merged**